File: src/main/java/com/gmail/borlandlp/minigamesdtools/creator/CreatorHub.java
package com.gmail.borlandlp.minigamesdtools.creator;

import com.gmail.borlandlp.minigamesdtools.Debug;

import java.util.ArrayList;
import java.util.Hashtable;
import java.util.List;
import java.util.Map;

public abstract class CreatorHub implements ICreatorHub {
    private Map<String, Creator> creators = new Hashtable<>();
    private Map<String, String> route = new Hashtable<>();

    /*
     * TODO: перевести на английский
     * Регистрирует создателя объектов.
     * */
    @Override
    public void registerCreator(Creator creator) throws Exception {
        CreatorInfo annotation = creator.getClass().getDeclaredAnnotation(CreatorInfo.class);
        if(annotation != null) {
            if(this.creators.containsKey(annotation.creatorId())) {
                StringBuilder msg = (new StringBuilder())
                        .append("Creator with ID:").append(annotation.creatorId())
                        .append(" is already registered for class:").append(this.creators.get(annotation.creatorId()).getClass().getName())
                        .append(" class trying to register:").append(creator.getClass().getName());
                throw new Exception(msg.toString());
            }

            this.creators.put(annotation.creatorId(), creator);
            Debug.print(Debug.LEVEL.NOTICE, "Register #class:" + creator.getClass().getSimpleName() + " with ID:" + annotation.creatorId());
        } else {
            throw new Exception("Creator must be have ID!");
        }
    }

    @Override
    public void unregisterCreator(Creator creator) {
        //this.route.keySet().parallelStream().filter(n -> this.routeId2Creator(n) == creator).forEach(n -> this.route.remove(n));
        this.creators.keySet().parallelStream().filter(n -> this.creators.get(n) == creator).forEach(n -> this.creators.remove(n));
    }

    @Override
    public void unregisterCreator(String creatorID) {
        if(this.creators.containsKey(creatorID)) {
            return;
        }

        this.unregisterCreator(this.creators.get(creatorID));
    }

    /*
     * TODO: перевести на английский
     * Регистрирует перенаправление с ID/псевдонима собираемого объекта(Фигурирует обычно в конфигах),
     * на конкретного создателя(ID указывается через аннотацию @CreatorInfo к создателю),
     * Который мог бы собрать этот объект.
     * */
    @Override
    public void registerRouteId2Creator(String routeId, String creatorID) throws Exception {
        if(!this.creators.containsKey(creatorID)) {
            throw new Exception("Cant find creator ID:" + creatorID + ", current itemID:" + routeId);
        } else {
            this.route.put(routeId, creatorID);
            Debug.print(Debug.LEVEL.NOTICE, "Register @route itemID:" + routeId + " for creatorID:" + creatorID);
        }
    }

    public Creator routeId2Creator(String Id) {
        return this.creators.get(this.route.get(Id));
    }

    @Override
    public void unregisterRouteId2Creator(String itemID) {
        this.route.remove(itemID);
    }

    /*
     * TODO: перевести на английский
     * Базовый метод создания объекта
     * */
    @Override
    public Object create(String itemID, AbstractDataProvider dataProvider) throws Exception {
        if(!this.containsRouteId2Creator(itemID)) {
            throw new Exception("itemID:" + itemID + " isnt registered routeID!");
        }

        Creator creator = this.routeId2Creator(itemID);
        List<String> reqFields = creator.getDataProviderRequiredFields();
        if(reqFields != null) {
            for (String reqField : creator.getDataProviderRequiredFields()) {
                if(!dataProvider.contains(reqField)) {
                    throw new Exception("Invalid DataProvider. Missing field '" + reqField + "'. itemID:" + itemID);
                }
            }
        }

        return creator.create(itemID, dataProvider);
    }

    @Override
    public boolean containsCreator(String ID) {
        return this.creators.containsKey(ID);
    }

    @Override
    public boolean containsCreator(Creator creator) {
        return this.creators.containsValue(creator);
    }

    @Override
    public boolean containsRouteId2Creator(String itemId) {
        return this.route.containsKey(itemId);
    }
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/creator/CreatorHub.kt
package com.gmail.borlandlp.minigamesdtools.creator

import com.gmail.borlandlp.minigamesdtools.Debug
import com.gmail.borlandlp.minigamesdtools.Debug.print
import java.util.*

abstract class CreatorHub : ICreatorHub {
    private val creators: MutableMap<String, Creator?> =
        Hashtable()
    private val route: MutableMap<String, String?> =
        Hashtable()

    /*
     * TODO: перевести на английский
     * Регистрирует создателя объектов.
     * */
    @Throws(Exception::class)
    override fun registerCreator(factory: Creator) {
        val annotation = factory.javaClass.getDeclaredAnnotation(CreatorInfo::class.java)
        if (annotation != null) {
            if (creators.containsKey(annotation.creatorId)) {//append(this.creators.get(annotation.creatorId()).getClass().getName())
                val msg = StringBuilder()
                    .append("Creator with ID:").append(annotation.creatorId)
                    .append(" is already registered for class:")
                    .append(creators[annotation.creatorId]!!::class.java.name)
                    .append(" class trying to register:").append(factory.javaClass.name)
                throw Exception(msg.toString())
            }
            creators[annotation.creatorId] = factory
            print(
                Debug.LEVEL.NOTICE,
                "Register #class:" + factory.javaClass.simpleName + " with ID:" + annotation.creatorId
            )
        } else {
            throw Exception("Creator must be have ID!")
        }
    }

    override fun unregisterCreator(creator: Creator) { //this.route.keySet().parallelStream().filter(n -> this.routeId2Creator(n) == creator).forEach(n -> this.route.remove(n));
        creators.keys.parallelStream()
            .filter { n: String? -> creators[n] === creator }
            .forEach { n: String? -> creators.remove(n) }
    }

    override fun unregisterCreator(creatorID: String) {
        if (creators.containsKey(creatorID)) {
            return
        }
        this.unregisterCreator(creators[creatorID]!!)
    }

    /*
     * TODO: перевести на английский
     * Регистрирует перенаправление с ID/псевдонима собираемого объекта(Фигурирует обычно в конфигах),
     * на конкретного создателя(ID указывается через аннотацию @CreatorInfo к создателю),
     * Который мог бы собрать этот объект.
     * */
    @Throws(Exception::class)
    override fun registerRouteId2Creator(itemID: String, creator: String) {
        if (!creators.containsKey(creator)) {
            throw Exception("Cant find creator ID:$creator, current itemID:$itemID")
        } else {
            route[itemID] = creator
            print(
                Debug.LEVEL.NOTICE,
                "Register @route itemID:$itemID for creatorID:$creator"
            )
        }
    }

    fun routeId2Creator(Id: String): Creator {
        return creators[route[Id]]!!
    }

    override fun unregisterRouteId2Creator(itemID: String) {
        route.remove(itemID)
    }

    /*
     * TODO: перевести на английский
     * Базовый метод создания объекта
     * */
    @Throws(Exception::class)
    override fun create(itemID: String, dataProvider: AbstractDataProvider): Any {
        if (!containsRouteId2Creator(itemID)) {
            throw Exception("itemID:$itemID isnt registered routeID!")
        }
        val creator = routeId2Creator(itemID)
        for (reqField in creator.dataProviderRequiredFields) {
            if (!dataProvider.contains(reqField)) {
                throw Exception("Invalid DataProvider. Missing field '$reqField'. itemID:$itemID")
            }
        }

        return creator.create(itemID, dataProvider)
    }

    override fun containsCreator(ID: String): Boolean {
        return creators.containsKey(ID)
    }

    override fun containsCreator(creator: Creator): Boolean {
        return creators.containsValue(creator)
    }

    override fun containsRouteId2Creator(itemId: String): Boolean {
        return route.containsKey(itemId)
    }
}